src/GaugeInput.js

```diff
@@ -0,0 +1,220 @@
+    width: '100%'
+  },
+  textField: {
+    width: '60%',
+    marginRight: 15
+  },
+  expansionPanelDetails: {
+    display: 'flex',
+    flexDirection: 'column'
+  }
+});
+
+class GaugeInput extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      id: `${this.props.id}`,
+      index: `${this.props.index}`,
+      prob: [],
+      success: false,
+      error: false,
+      warning: false,
+      gauge: this.props.gauge
+    }
+  }
+
+  handleOnChange = event => {
+    console.log('Click');
+    console.log(event.target.value);
+  };
+
+  open = (result) => {
+    console.log(result)
+    if (result.errors) {
+      if (this.props.pass === "") {
+        this.setState({warning: true})
+      } else {
+        this.setState({error: true})
+      }
+    } else {
+      this.setState({success: true})
+    }
+  }
+
+  close = (event, reason) => {
+    if (reason === 'clickaway') {
+      return;
+    }
+    if (this.state.error) {
+      this.setState({error: false})
+    } else if (this.state.warning){
+      this.setState({warning: false})
+    } else {
+      this.setState({success: false})
+    }
+  }
+
+  render() {
+    const { classes } = this.props;
+    let { gauge, prob } = this.state;
+    let text = []
+
+    if (gauge.messages) {
+      gauge.messages.map((m => {
+        prob.push(m.probability)
+        text.push(m.text)
+      }))
+    }
+
+    const updateProb = event => {
+      prob[event.target.id][event.target.getAttribute('bin')] = event.target.value
+    }
+
+    const updateText = event => {
+      text[event.target.id] = event.target.value
+    }
+
+    const addMessage = () => {
+      api.post(`glyphs/${this.state.id}/gauges/${this.state.index}/messages`,
+        {"pass" : this.props.pass, "text" : 'Default message', "probability" : [0,0,0,0,0]}
+      ).then(result => {console.log(result)})
+
+      if (!gauge.messages) {
+        gauge[ "messages" ] = [];
+      }
+      gauge.messages.push({"text" : 'Default gauge message', "probability" : [0,0,0,0,0]})
+      this.forceUpdate();
+    }
+
+    const updateMessage = event => {
+      const index = event.target.id
+      api.post(`glyphs/${this.state.id}/gauges/${this.state.index}/messages/${event.target.id + 1}`,
+        {"pass" : this.props.pass, "text" : text[index], "probability" : prob[index]}
+      ).then(result => {this.open(result)})
+    }
+
+    if (gauge.messages) {
+      console.log('Messages attached')
+      return (
+        <div>
+          <ExpansionPanel style={{width: '50%', border: '1px solid rgba(0, 0, 0, .125)', marginTop: 10}}>
+            <ExpansionPanelSummary
+              expandIcon={<ExpandMoreIcon />}
+              aria-controls="panel1a-content"
+              id="panel1a-header">
+              Gauge {this.state.index}
+            </ExpansionPanelSummary>
+            <ExpansionPanelDetails className={classes.expansionPanelDetails}>
+              {gauge.messages.map((m, index) =>
```

Comment:
There is too much repetitive code in this section.  The upper-level if `if (gauge.messages)` can be interweaved into this logic.  For example, here you can write `{gauge.messages && gauge.messages.map`